### PR TITLE
Bitbucket Support for MLFlow

### PR DESCRIPTION
**Pull request by rohankamath**

### CODE REVIEW

File: mlflow/server/js/src/utils/Utils.js [hunk 1]
```diff
@@ -6,12 +6,12 @@ import laptopSvg from '../static/laptop.svg';
 import projectSvg from '../static/project.svg';
```
Comment (mateiz) [inline mlflow/server/js/src/utils/Utils.js]: It looks like your IDE reformatted this whole file unfortunately. Can you reduce the diff to just your changes without changing the other lines?

File: mlflow/server/js/src/utils/Utils.js [hunk 2]
```diff
@@ -114,21 +114,32 @@ class Utils {
...
+    /**
+     * Renders the source name and entry point into an HTML element. Used for display.
+     * @param run MlflowMessages.RunInfo
+     * @param tags Object containing tag key value pairs.
+     */
   static renderSource(run, tags) {
     let res = Utils.formatSource(run);
     if (run.source_type === "PROJECT") {
       const match = run.source_name.match(Utils.getGitHubRegex());
+      const match1 = run.source_name.match(Utils.getBitBucketRegex());
```
Comment (mateiz) [inline mlflow/server/js/src/utils/Utils.js]: Give these slightly more readable names, such as githubMatch and bitBucketMatch.

File: mlflow/server/js/src/utils/Utils.test.js [hunk 3]
```diff
@@ -1,9 +1,7 @@
 import Utils from './Utils';
 import { RunInfo } from '../sdk/MlflowMessages';
-import React from 'react';
```
Comment (mateiz) [inline mlflow/server/js/src/utils/Utils.test.js]: Why was this changed?

File: mlflow/server/js/src/utils/Utils.js (hunk 1 above)
Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.js]: Yes. I'll look into this and do the needful. Thanks

File: mlflow/server/js/src/utils/Utils.test.js (hunk 3 above)
Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.test.js]: 4:8  error  'React' is defined but never used  no-unused-vars

This is the error that I receive when it's there.

File: mlflow/server/js/src/utils/Utils.js (hunk 2 above)
Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.js]: Done

File: mlflow/server/js/src/utils/Utils.test.js (hunk 3 above)
Comment (mateiz) [reply mlflow/server/js/src/utils/Utils.test.js]: That seems to be because you deleted a line I the test below. Why is that?

**Comment by mateiz** [comment]
It looks like this still has a lot of whitespace differences; can you reduce the diff to just the lines you need to add? You can see them in the Files view.

File: mlflow/server/js/src/utils/Utils.test.js (hunk 3 above)
Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.test.js]: Added the deleted lines. Fixed that.

**Comment by rohankamath1** [comment]
> It looks like this still has a lot of whitespace differences; can you reduce the diff to just the lines you need to add? You can see them in the Files view.

Done

**Comment by rohankamath1** [comment]
Hey! Can you check and let me know if any other changes are required? Thanks a lot! @mateiz

File: mlflow/server/js/src/utils/Utils.js [hunk 4]
```diff
@@ -122,11 +125,11 @@ class Utils {
...
-      const match = run.source_name.match(Utils.getGitHubRegex());
-      if (match) {
-        let url = "https://github.com/" + match[1] + "/" + match[2].replace(/.git/, '');
-        if (match[3]) {
-          url = url + "/tree/master/" + match[3];
+      const GitHubMatch = run.source_name.match(Utils.getGitHubRegex());
+      if (GitHubMatch) {
+        let url = "https://github.com/" + GitHubMatch[1] + "/" + GitHubMatch[2].replace(/.git/, '');
+        if (GitHubMatch[3]) {
+          url = url + "/tree/master/" + GitHubMatch[3];
```
Comment (mateiz) [inline mlflow/server/js/src/utils/Utils.js]: Shouldn't this code be updated to support BitBucket too?

File: mlflow/server/js/src/utils/Utils.js [hunk 5]
```diff
@@ -209,10 +212,15 @@ class Utils {
     if (run.source_version) {
       const versionString = shortVersion ? run.source_version.substring(0, 6) : run.source_version;
       if (run.source_type === "PROJECT") {
-        const match = run.source_name.match(Utils.getGitHubRegex());
-        if (match) {
-          const url = ("https://github.com/" + match[1] + "/" + match[2].replace(/.git/, '') +
-                     "/tree/" + run.source_version) + "/" + match[3];
+        const GitHubMatch = run.source_name.match(Utils.getGitHubRegex());
+        const BitBucketMatch = run.source_name.match(Utils.getBitBucketRegex());
```
Comment (mateiz) [inline mlflow/server/js/src/utils/Utils.js]: Nit: You should capitalize these as gitHubMatch and bitBucketMatch (camel case) to match the code style in the rest of the codebase.

File: mlflow/server/js/src/utils/Utils.js [hunk 6]
```diff
@@ -122,11 +125,11 @@ class Utils {
   static renderSource(run, tags) {
     let res = Utils.formatSource(run);
     if (run.source_type === "PROJECT") {
-      const match = run.source_name.match(Utils.getGitHubRegex());
-      if (match) {
-        let url = "https://github.com/" + match[1] + "/" + match[2].replace(/.git/, '');
-        if (match[3]) {
-          url = url + "/tree/master/" + match[3];
+      const GitHubMatch = run.source_name.match(Utils.getGitHubRegex());
```
Comment (mateiz) [inline mlflow/server/js/src/utils/Utils.js]: should be capitalized as gitHubMatch

Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.js]: Done

File: mlflow/server/js/src/utils/Utils.js (hunk 5 above)
Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.js]: Done

File: mlflow/server/js/src/utils/Utils.js (hunk 4 above)
Comment (rohankamath1) [reply mlflow/server/js/src/utils/Utils.js]: Before I update that code,  I am facing this issue which I'm not able to solve. 

Line 232: static is a reserved word in strict mode

Kindly help. Thanks

**Comment by mateiz** [comment]
Sorry for the delay responding. It looks like this is still failing tests with a JS error (https://travis-ci.org/mlflow/mlflow/jobs/485871764). If you can't fix it, let me know and I can try to take over the patch and fix it.

**Comment by rohankamath1** [comment]
> Sorry for the delay responding. It looks like this is still failing tests with a JS error (https://travis-ci.org/mlflow/mlflow/jobs/485871764). If you can't fix it, let me know and I can try to take over the patch and fix it.

Sure! You can take over the patch and fix it. That would be great.

**Comment by mateiz** [comment]
FYI, this actually got fixed along with GitLab in #901 so I'm going to close this PR. Thanks for your work on it and for raising this issue! It should be out in the next release.
